mnefun/_mnefun.py

```diff
@@ -2391,6 +2391,194 @@ def plot_reconstruction(evoked, origin=(0., 0., 0.04)):
+    nepochs, nchan, buflen = alldata.shape
+    sfreq = epochs.info['sfreq']
+    t = np.linspace(0, buflen/sfreq, endpoint=False, num=buflen)
+    (cfreqs, _, _, _, _) = _get_hpi_info(epochs.info)
+    ncoils = len(cfreqs)
+    linefreq = epochs.info['line_freq']
+    linefreqs = (np.arange(n_lineharm+1)+1) * linefreq
+
+    # gradiometer and magmeter indices
+    pick_mag = pick_types(epochs.info, meg='mag')
+    pick_grad = pick_types(epochs.info, meg='grad')
+
+    # create linear model
+    model = np.c_[t, np.ones(t.shape)]  # model slope and DC
+    # add sine and cosine term for each freq
+    for f in list(linefreqs)+list(cfreqs):
+        model = np.c_[model, np.cos(2*np.pi*f*t), np.sin(2*np.pi*f*t)]
+    inv_model = np.linalg.pinv(model)
+
+    # loop through epochs
+    snr_avg_grad = np.zeros([ncoils, nepochs])
+    snr_avg_mag = np.zeros([ncoils, nepochs])
+    resid_vars = np.zeros([nchan, nepochs])
+
+    for epn in range(nepochs):
+        epdata = alldata[epn, :, :].transpose()
+        coeffs = np.dot(inv_model, epdata)
+        coeffs_hpi = coeffs[2 + 2*len(linefreqs):]
+        resid_vars[:, epn] = np.var(epdata - np.dot(model, coeffs), 0)
+        # get total hpi amplitudes by combining sine and cosine terms
+        hpi_amps = np.sqrt(coeffs_hpi[0::2, :]**2 + coeffs_hpi[1::2, :]**2)
+        # divide average HPI power by average variance
+        snr_avg_grad[:, epn] = np.divide((hpi_amps**2/2)[:, pick_grad].mean(1),
+                                         resid_vars[pick_grad, epn].mean())
+        snr_avg_mag[:, epn] = np.divide((hpi_amps**2/2)[:, pick_mag].mean(1),
+                                        resid_vars[pick_mag, epn].mean())
+    if hpi_coil == 'median':
+        fun = np.median
+    elif hpi_coil == 'best':
+        fun = np.max
+    elif hpi_coil == 'worst':
+        fun = np.min
+    else:
+        raise ValueError('Invalid HPI coil selection parameter')
+
+    if channels == 'grad':
+        snr = fun(snr_avg_grad, axis=0)
+    elif channels == 'mag':
+        snr = fun(snr_avg_mag, axis=0)
+    return snr
+
+
+def _weigh_epochs(epochs, weights):
+    """ Weigh epochs in mne.Epochs instance (= elementwise multiply by
+    weights vector). """
+    epochs.load_data()
+    weights = np.array(weights)  # will accept either list or numpy array
+    n_epochs = len(epochs)
+    if not len(weights) == n_epochs:
+        raise ValueError('Need as many weights as epochs')
+    # reshape for broadcasting
+    w_ = weights.squeeze()[:, np.newaxis, np.newaxis]
+    # normalize by n_epochs/sum(weights), so that averaging
+    # will yield a correct result
+    epochs._data *= n_epochs * w_ / np.sum(w_)
+
+
+def chpi_weighted_average(raw_snr, raw_epochs=None, reject=False, flat=False,
+                          nharm=None, epoch_start=None, epoch_end=None,
+                          stim_channel=None, mask=None):
+    """ Average epochs weighted by SNR of continuous HPI (cHPI).
```

Comment:
I've been using this with the typical MaxFilter workflow, i.e. I have the original raw file with cHPI signals intact, and a MaxFiltered version of the same file (tSSS and movement compensation done). Then the idea is to pick the cHPI signals from the original file, and the filtered epochs from the MaxFiltered file, so you need two separate raw file arguments.

If I get @Eric89GXL correctly, you would like to feed already epoched (filtered) data to the method, and it would extract the cHPI epochs from raw data accordingly. I guess the need for that is somehow related to your specific epoching procedures, because otherwise it's simpler just to feed the two Raw instances to it, right?